### PR TITLE
Add support for per-PDS event rate limits in BGS

### PR DESCRIPTION
**Pull request by ericvolp12**
This change adds support for runtime-configurable event rate limits on a per-PDS basis.

One new admin endpoint has been added to change the event rate limit of a PDS on-demand.

The PDS listing admin endpoint now returns the current rate limit for each PDS and their available tokens (sort of a proxy to events per second, may need some more interpretation).

- [x] Add a flow for changing PDS rate-limit through the admin dashboard.

### CODE REVIEW

**Comment by ericvolp12** [comment]
Now supports changing the rate limit from the Admin UI

CleanShot 2023-08-21 at 20 56 08

**Comment by ericvolp12** [comment]
Now with support for configurable rate limits for Crawling PDSs so you don't DOS them with GetCheckouts on connect.

CleanShot 2023-08-21 at 21 44 14

File: bgs/fedmgr.go [hunk 1]
```diff
@@ -226,6 +266,17 @@ func (s *Slurper) RestartAll() error {
 			cancel: cancel,
 		}
 		s.active[pds.Host] = &sub
+		// Check if we've already got a limiter for this PDS
+		s.LimitMux.RLock()
```
Comment (whyrusleeping) [inline bgs/fedmgr.go:270]: id pull this logic into a separate function

File: bgs/fedmgr.go [hunk 2]
```diff
@@ -389,15 +440,22 @@ func (s *Slurper) handleConnection(ctx context.Context, host *models.PDS, con *w
 		},
 	}
 
+	s.LimitMux.RLock()
```
Comment (whyrusleeping) [inline bgs/fedmgr.go:443]: a general code cleanliness thing im a fan of is if you have a lock in the middle of your function, it probably wants to be in its own function

File: bgs/admin.go [hunk 3]
```diff
@@ -91,10 +92,17 @@ func (bgs *BGS) handleAdminGetUpstreamConns(e echo.Context) error {
...
+}
+
 type enrichedPDS struct {
 	models.PDS
-	HasActiveConnection    bool   `json:"HasActiveConnection"`
-	EventsSeenSinceStartup uint64 `json:"EventsSeenSinceStartup"`
+	HasActiveConnection    bool      `json:"HasActiveConnection"`
+	EventsSeenSinceStartup uint64    `json:"EventsSeenSinceStartup"`
+	IngestRateLimit        rateLimit `json:"IngestRateLimit"`
+	CrawlRateLimit         rateLimit `json:"CrawlRateLimit"`
```
[the remark this answers is not in the document]
Comment (ericvolp12) [reply bgs/admin.go:105]: It is now but I kinda want the rateLimits to have the same shape in this response

**Review by whyrusleeping — APPROVED** [review]
field name conflict aside, this LGTM

File: bgs/admin.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (ericvolp12) [reply bgs/admin.go]: Resolved

**Pull request merged**